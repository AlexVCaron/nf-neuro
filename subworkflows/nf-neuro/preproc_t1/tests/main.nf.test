nextflow_workflow {

    name "Test Subworkflow PREPROC_T1"
    script "../main.nf"
    workflow "PREPROC_T1"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/preproc_t1"

    tag "denoising/nlmeans"
    tag "preproc/n4"
    tag "image/resample"
    tag "betcrop/antsbet"
    tag "betcrop/synthbet"
    tag "image/cropvolume"

    tag "load_test_data"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "antsbet.zip" , "T1w.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("preproc_t1_antsbet_error") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assert workflow.failed
        }
    }

    test("preproc_t1_antsbet") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz"),
                    ]}
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel ->
                        channel.value.every{ item ->
                            item instanceof ArrayList ? item.get(1) : item } } }
            )
        }
    }

    test("preproc_t1_synthbet") {
        config "./nextflow_synthbet.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel ->
                        channel.value.every{ item ->
                            item instanceof ArrayList ? item.get(1) : item } } }
            )
        }
    }

    test("preproc_t1_quick") {
        config "./nextflow_quick.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["image_N4",
                                        "image_bet",
                                        "mask_bet"].contains(channel.key)
                        ? channel.value.size() == 0
                        : channel.value.every{ item ->
                            item instanceof ArrayList ? item.get(1) : item } } }
            )
        }
    }

    test("preproc_t1_skip_all") {
        config "./nextflow_skip_all.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["image_nlmeans",
                                        "image_N4",
                                        "image_resample",
                                        "image_bet",
                                        "mask_bet",
                                        "crop_box"].contains(channel.key)
                        ? channel.value.size() == 0
                        : channel.value.every{ item ->
                            item instanceof ArrayList ? item.get(1) : item } } }
            )
        }
    }
}
