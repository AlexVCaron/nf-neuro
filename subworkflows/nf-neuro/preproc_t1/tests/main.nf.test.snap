{
    "preproc_t1_quick": {
        "content": [
            {
                "image_nlmeans": [
                    [
                        {
                            "id": "test"
                        },
                        "test_denoised.nii.gz"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test_resampled.nii.gz"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b3052ba8e0fbaee41f2064801434f38b",
                    "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                    "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-18T04:02:31.202475083"
    },
    "preproc_t1_synthbet": {
        "content": [
            {
                "crop_box": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped_bbox.pkl"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__image_n4.nii.gz"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__bet_image.nii.gz"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test"
                        },
                        "test_denoised.nii.gz"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test_resampled.nii.gz"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__brain_mask.nii.gz"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cropped.nii.gz"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,318cabe934be45528a25f52083d9c90d",
                    "versions.yml:md5,7aad91b5043d2dc8e8fc69ffbbe6f657",
                    "versions.yml:md5,b3052ba8e0fbaee41f2064801434f38b",
                    "versions.yml:md5,be3dbb0ac2589ad263d583018f339102",
                    "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                    "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-18T03:54:59.425262637"
    },
    "preproc_t1_skip_all": {
        "content": [
            {
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "T1w.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-16T10:35:13.269602"
    },
    "preproc_t1_antsbet": {
        "content": [
            {
                "crop_box": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped_bbox.pkl"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__image_n4.nii.gz"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_bet.nii.gz"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test"
                        },
                        "test_denoised.nii.gz"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test_resampled.nii.gz"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_bet_mask.nii.gz"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cropped.nii.gz"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3bda0d9faae5e07d500a04a0119b766f",
                    "versions.yml:md5,7aad91b5043d2dc8e8fc69ffbbe6f657",
                    "versions.yml:md5,b3052ba8e0fbaee41f2064801434f38b",
                    "versions.yml:md5,be3dbb0ac2589ad263d583018f339102",
                    "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                    "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-18T03:52:58.329141544"
    }
}